extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp

```diff
@@ -0,0 +1,130 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "TestBase.h"
+#include "YamlConfiguration.h"
+#include "Utils.h"
+#include "IntegrationTestUtils.h"
+#include "ProcessGroupTestUtils.h"
+
+static core::YamlConfiguration config(nullptr, nullptr, nullptr, nullptr, std::make_shared<minifi::Configure>());
+
+TEST_CASE("Root process group is correctly parsed", "[YamlProcessGroupParser1]") {
+  auto pattern = Group("root")
+    .With({
+      Conn{"Conn1",
+           MaybeProc{"00000000-0000-0000-0000-000000000001", "Proc1"},
+           MaybeProc{"00000000-0000-0000-0000-000000000002", "Proc2"}},
+      Conn{"Conn2",
+           MaybeProc{"00000000-0000-0000-0000-000000000002", "Proc2"},
+           MaybeProc{"00000000-0000-0000-0000-000000000005", "Port1"}}
+    }).With({
+      Proc{"Proc1", "00000000-0000-0000-0000-000000000001"},
+      Proc{"Proc2", "00000000-0000-0000-0000-000000000002"}
+    }).With({
+      RPG{"RPG1", {
+        Proc{"Port1", "00000000-0000-0000-0000-000000000005"},
+        Proc{"Port2", "00000000-0000-0000-0000-000000000006"}
+      }},
+      RPG{"RPG2", {}}
+    });
+
+  auto root = config.getRootFromPayload(pattern.serialize().join("\n"));
+  verifyProcessGroup(*root, pattern);
+}
+
+TEST_CASE("Nested process group is correctly parsed", "[YamlProcessGroupParser2]") {
+  auto pattern = Group("root")
+    .With({Conn{"Conn1",
+                MaybeProc{"00000000-0000-0000-0000-000000000001", "Proc1"},
+                MaybeProc{"00000000-0000-0000-0000-000000000002", "Proc2"}}})
+    .With({Proc{"Proc1", "00000000-0000-0000-0000-000000000001"},
+           Proc{"Proc2", "00000000-0000-0000-0000-000000000002"}})
+    .With({
+      Group("Child1")
+      .With({Conn{"Child1_Conn1",
+                  MaybeProc{"00000000-0000-0000-0000-000000000005", "Port1"},
+                  MaybeProc{"00000000-0000-0000-0000-000000000007", "Child1_Proc2"}}})
+      .With({Proc{"Child1_Proc1", "00000000-0000-0000-0000-000000000006"},
+             Proc{"Child1_Proc2", "00000000-0000-0000-0000-000000000007"}})
+      .With({RPG{"Child1_RPG1", {
+        Proc{"Port1", "00000000-0000-0000-0000-000000000005"}}}})
+    });
+
+  auto root = config.getRootFromPayload(pattern.serialize().join("\n"));
+
+  verifyProcessGroup(*root, pattern);
+}
+
+TEST_CASE("Cannot connect processors from different groups", "[YamlProcessGroupParser3]") {
```

Comment:
The name of this test is misleading, as it verifies that you _can_ connect processors from different groups.

Reply:
this test was quite cryptic, I have changed `MaybeProc{"<id>", utils::nullopt}` to `UnresolvedProc{"<id>"}`, does this make it easier to understand?

Reply:
Yes, this helps.  I forgot that a null optional for the ID means "verify that this processor is not found".  It is still a bit tricky, but clearer.